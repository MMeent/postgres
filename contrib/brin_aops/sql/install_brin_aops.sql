CREATE EXTENSION brin_aops;

CREATE TABLE test (dataset int2[], padding text) with (fillfactor = 10);
select setseed(0);
INSERT INTO test
SELECT array(select distinct (random() * n/5)::int2 + n / 10 + (y % ((n + 16) / 17)) + 1
			 from generate_series(n, n + 100, 1) y
			 order by 1 limit 15
		   )::int2[] AS dataset,
	   repeat('a', 800) AS padding
FROM generate_series(1, 1000) n;

EXPLAIN (costs off, analyze, verbose off, buffers, timing off)
	SELECT count(*) FROM test WHERE dataset > '{1}'::int2[];
SELECT count(*) FROM test WHERE dataset > '{1}'::int2[];
EXPLAIN (costs off, analyze, verbose off, buffers, timing off)
	SELECT count(*) FROM test WHERE dataset && '{1}'::int2[];
SELECT count(*) FROM test WHERE dataset && '{1}'::int2[];
EXPLAIN (costs off, analyze, verbose off, buffers, timing off)
	SELECT count(*) FROM test WHERE dataset && '{null}'::int2[];
SELECT count(*) FROM test WHERE dataset && '{null}'::int2[];
EXPLAIN (costs off, analyze, verbose off, buffers, timing off)
	SELECT count(*) FROM test WHERE dataset @> '{}'::int2[];
SELECT count(*) FROM test WHERE dataset @> '{}'::int2[];
EXPLAIN (costs off, analyze, verbose off, buffers, timing off)
	SELECT count(*) FROM test WHERE dataset <@ '{1,2,3,4,5,6,7,8,9,10}'::int2[];
SELECT count(*) FROM test WHERE dataset <@ '{1,2,3,4,5,6,7,8,9,10}'::int2[];

CREATE INDEX brin_minmax_int2
    ON test
        USING brin (dataset int2_minmax_aops)
		WITH (pages_per_range = 2);

SET enable_seqscan = off;

EXPLAIN (costs off, analyze, verbose off, buffers, timing off)
	SELECT count(*) FROM test WHERE dataset > '{1}'::int2[];
SELECT count(*) FROM test WHERE dataset > '{1}'::int2[];
EXPLAIN (costs off, analyze, verbose off, buffers, timing off)
	SELECT count(*) FROM test WHERE dataset && '{1}'::int2[];
SELECT count(*) FROM test WHERE dataset && '{1}'::int2[];
EXPLAIN (costs off, analyze, verbose off, buffers, timing off)
	SELECT count(*) FROM test WHERE dataset && '{null}'::int2[];
SELECT count(*) FROM test WHERE dataset && '{null}'::int2[];
EXPLAIN (costs off, analyze, verbose off, buffers, timing off)
	SELECT count(*) FROM test WHERE dataset @> '{}'::int2[];
SELECT count(*) FROM test WHERE dataset @> '{}'::int2[];
EXPLAIN (costs off, analyze, verbose off, buffers, timing off)
	SELECT count(*) FROM test WHERE dataset <@ '{1,2,3,4,5,6,7,8,9,10}'::int2[];
SELECT count(*) FROM test WHERE dataset <@ '{1,2,3,4,5,6,7,8,9,10}'::int2[];
