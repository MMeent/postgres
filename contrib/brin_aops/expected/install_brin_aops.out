CREATE EXTENSION brin_aops;
CREATE TABLE test (dataset int2[], padding text) with (fillfactor = 10);
select setseed(0);
 setseed
---------

(1 row)

INSERT INTO test
SELECT array(select distinct (random() * n/5)::int2 + n / 10 + (y % ((n + 16) / 17)) + 1
			 from generate_series(n, n + 100, 1) y
			 order by 1 limit 15
		   )::int2[] AS dataset,
	   repeat('a', 800) AS padding
FROM generate_series(1, 1000) n;
EXPLAIN (costs off, analyze, verbose off, buffers, timing off)
	SELECT count(*) FROM test WHERE dataset > '{1}'::int2[];
                    QUERY PLAN
--------------------------------------------------
 Aggregate (actual rows=1 loops=1)
   Buffers: shared hit=1031
   ->  Seq Scan on test (actual rows=998 loops=1)
         Filter: (dataset > '{1}'::smallint[])
         Rows Removed by Filter: 2
         Buffers: shared hit=1031
 Planning:
   Buffers: shared hit=10 read=2
 Planning Time: 0.091 ms
 Execution Time: 1.799 ms
(10 rows)

SELECT count(*) FROM test WHERE dataset > '{1}'::int2[];
 count
-------
   998
(1 row)

EXPLAIN (costs off, analyze, verbose off, buffers, timing off)
	SELECT count(*) FROM test WHERE dataset && '{1}'::int2[];
                   QUERY PLAN
------------------------------------------------
 Aggregate (actual rows=1 loops=1)
   Buffers: shared hit=1003
   ->  Seq Scan on test (actual rows=9 loops=1)
         Filter: (dataset && '{1}'::smallint[])
         Rows Removed by Filter: 991
         Buffers: shared hit=1003
 Planning Time: 0.021 ms
 Execution Time: 0.996 ms
(8 rows)

SELECT count(*) FROM test WHERE dataset && '{1}'::int2[];
 count
-------
     9
(1 row)

EXPLAIN (costs off, analyze, verbose off, buffers, timing off)
	SELECT count(*) FROM test WHERE dataset && '{null}'::int2[];
                    QUERY PLAN
---------------------------------------------------
 Aggregate (actual rows=1 loops=1)
   Buffers: shared hit=1000
   ->  Seq Scan on test (actual rows=0 loops=1)
         Filter: (dataset && '{NULL}'::smallint[])
         Rows Removed by Filter: 1000
         Buffers: shared hit=1000
 Planning Time: 0.024 ms
 Execution Time: 0.916 ms
(8 rows)

SELECT count(*) FROM test WHERE dataset && '{null}'::int2[];
 count
-------
     0
(1 row)

EXPLAIN (costs off, analyze, verbose off, buffers, timing off)
	SELECT count(*) FROM test WHERE dataset @> '{}'::int2[];
                    QUERY PLAN
---------------------------------------------------
 Aggregate (actual rows=1 loops=1)
   Buffers: shared hit=1000
   ->  Seq Scan on test (actual rows=1000 loops=1)
         Filter: (dataset @> '{}'::smallint[])
         Buffers: shared hit=1000
 Planning Time: 0.019 ms
 Execution Time: 0.770 ms
(7 rows)

SELECT count(*) FROM test WHERE dataset @> '{}'::int2[];
 count
-------
  1000
(1 row)

EXPLAIN (costs off, analyze, verbose off, buffers, timing off)
	SELECT count(*) FROM test WHERE dataset <@ '{1,2,3,4,5,6,7,8,9,10}'::int2[];
                            QUERY PLAN
-------------------------------------------------------------------
 Aggregate (actual rows=1 loops=1)
   Buffers: shared hit=1000
   ->  Seq Scan on test (actual rows=29 loops=1)
         Filter: (dataset <@ '{1,2,3,4,5,6,7,8,9,10}'::smallint[])
         Rows Removed by Filter: 971
         Buffers: shared hit=1000
 Planning Time: 0.119 ms
 Execution Time: 1.270 ms
(8 rows)

SELECT count(*) FROM test WHERE dataset <@ '{1,2,3,4,5,6,7,8,9,10}'::int2[];
 count
-------
    29
(1 row)

EXPLAIN (costs off, analyze, verbose off, buffers, timing off)
	SELECT count(*) FROM test WHERE dataset = '{100}'::int2[];
                   QUERY PLAN
-------------------------------------------------
 Aggregate (actual rows=1 loops=1)
   Buffers: shared hit=1000
   ->  Seq Scan on test (actual rows=0 loops=1)
         Filter: (dataset = '{100}'::smallint[])
         Rows Removed by Filter: 1000
         Buffers: shared hit=1000
 Planning:
   Buffers: shared hit=59
 Planning Time: 0.195 ms
 Execution Time: 0.861 ms
(10 rows)

SELECT count(*) FROM test WHERE dataset = '{100}'::int2[];
 count
-------
     0
(1 row)

EXPLAIN (costs off, analyze, verbose off, buffers, timing off)
	SELECT count(*) FROM test WHERE dataset = '{68,73,74,77,79,80,84,85,88,89,90,94,99,100,101}'::int2[];
                                         QUERY PLAN
--------------------------------------------------------------------------------------------
 Aggregate (actual rows=1 loops=1)
   Buffers: shared hit=1000
   ->  Seq Scan on test (actual rows=1 loops=1)
         Filter: (dataset = '{68,73,74,77,79,80,84,85,88,89,90,94,99,100,101}'::smallint[])
         Rows Removed by Filter: 999
         Buffers: shared hit=1000
 Planning Time: 0.069 ms
 Execution Time: 0.821 ms
(8 rows)

SELECT count(*) FROM test WHERE dataset = '{68,73,74,77,79,80,84,85,88,89,90,94,99,100,101}'::int2[];
 count
-------
     1
(1 row)

CREATE INDEX brin_minmax_int2
    ON test
        USING brin (dataset int2_minmax_aops)
		WITH (pages_per_range = 2);
SET enable_seqscan = off;
EXPLAIN (costs off, analyze, verbose off, buffers, timing off)
	SELECT count(*) FROM test WHERE dataset > '{1}'::int2[];
                                  QUERY PLAN
------------------------------------------------------------------------------
 Aggregate (actual rows=1 loops=1)
   Buffers: shared hit=1007
   ->  Bitmap Heap Scan on test (actual rows=998 loops=1)
         Recheck Cond: (dataset > '{1}'::smallint[])
         Heap Blocks: lossy=998
         Buffers: shared hit=1007
         ->  Bitmap Index Scan on brin_minmax_int2 (actual rows=9980 loops=1)
               Index Cond: (dataset > '{1}'::smallint[])
               Buffers: shared hit=9
 Planning:
   Buffers: shared hit=21
 Planning Time: 0.130 ms
 Execution Time: 1.157 ms
(13 rows)

SELECT count(*) FROM test WHERE dataset > '{1}'::int2[];
 count
-------
   998
(1 row)

EXPLAIN (costs off, analyze, verbose off, buffers, timing off)
	SELECT count(*) FROM test WHERE dataset && '{1}'::int2[];
                                 QUERY PLAN
-----------------------------------------------------------------------------
 Aggregate (actual rows=1 loops=1)
   Buffers: shared hit=16
   ->  Bitmap Heap Scan on test (actual rows=9 loops=1)
         Recheck Cond: (dataset && '{1}'::smallint[])
         Rows Removed by Index Recheck: 1
         Heap Blocks: lossy=10
         Buffers: shared hit=16
         ->  Bitmap Index Scan on brin_minmax_int2 (actual rows=100 loops=1)
               Index Cond: (dataset && '{1}'::smallint[])
               Buffers: shared hit=6
 Planning:
   Buffers: shared hit=4
 Planning Time: 0.039 ms
 Execution Time: 0.283 ms
(14 rows)

SELECT count(*) FROM test WHERE dataset && '{1}'::int2[];
 count
-------
     9
(1 row)

EXPLAIN (costs off, analyze, verbose off, buffers, timing off)
	SELECT count(*) FROM test WHERE dataset && '{null}'::int2[];
                                QUERY PLAN
---------------------------------------------------------------------------
 Aggregate (actual rows=1 loops=1)
   Buffers: shared hit=3
   ->  Bitmap Heap Scan on test (actual rows=0 loops=1)
         Recheck Cond: (dataset && '{NULL}'::smallint[])
         Buffers: shared hit=3
         ->  Bitmap Index Scan on brin_minmax_int2 (actual rows=0 loops=1)
               Index Cond: (dataset && '{NULL}'::smallint[])
               Buffers: shared hit=3
 Planning:
   Buffers: shared hit=1
 Planning Time: 0.026 ms
 Execution Time: 0.609 ms
(12 rows)

SELECT count(*) FROM test WHERE dataset && '{null}'::int2[];
 count
-------
     0
(1 row)

EXPLAIN (costs off, analyze, verbose off, buffers, timing off)
	SELECT count(*) FROM test WHERE dataset @> '{}'::int2[];
                                  QUERY PLAN
-------------------------------------------------------------------------------
 Aggregate (actual rows=1 loops=1)
   Buffers: shared hit=1003
   ->  Bitmap Heap Scan on test (actual rows=1000 loops=1)
         Recheck Cond: (dataset @> '{}'::smallint[])
         Heap Blocks: lossy=1000
         Buffers: shared hit=1003
         ->  Bitmap Index Scan on brin_minmax_int2 (actual rows=10000 loops=1)
               Index Cond: (dataset @> '{}'::smallint[])
               Buffers: shared hit=3
 Planning:
   Buffers: shared hit=4
 Planning Time: 0.037 ms
 Execution Time: 1.322 ms
(13 rows)

SELECT count(*) FROM test WHERE dataset @> '{}'::int2[];
 count
-------
  1000
(1 row)

EXPLAIN (costs off, analyze, verbose off, buffers, timing off)
	SELECT count(*) FROM test WHERE dataset <@ '{1,2,3,4,5,6,7,8,9,10}'::int2[];
                                 QUERY PLAN
-----------------------------------------------------------------------------
 Aggregate (actual rows=1 loops=1)
   Buffers: shared hit=101
   ->  Bitmap Heap Scan on test (actual rows=29 loops=1)
         Recheck Cond: (dataset <@ '{1,2,3,4,5,6,7,8,9,10}'::smallint[])
         Rows Removed by Index Recheck: 69
         Heap Blocks: lossy=98
         Buffers: shared hit=101
         ->  Bitmap Index Scan on brin_minmax_int2 (actual rows=980 loops=1)
               Index Cond: (dataset <@ '{1,2,3,4,5,6,7,8,9,10}'::smallint[])
               Buffers: shared hit=3
 Planning:
   Buffers: shared hit=4
 Planning Time: 0.033 ms
 Execution Time: 0.534 ms
(14 rows)

SELECT count(*) FROM test WHERE dataset <@ '{1,2,3,4,5,6,7,8,9,10}'::int2[];
 count
-------
    29
(1 row)

EXPLAIN (costs off, analyze, verbose off, buffers, timing off)
	SELECT count(*) FROM test WHERE dataset = '{100}'::int2[];
                                  QUERY PLAN
------------------------------------------------------------------------------
 Aggregate (actual rows=1 loops=1)
   Buffers: shared hit=349
   ->  Bitmap Heap Scan on test (actual rows=0 loops=1)
         Recheck Cond: (dataset = '{100}'::smallint[])
         Rows Removed by Index Recheck: 346
         Heap Blocks: lossy=346
         Buffers: shared hit=349
         ->  Bitmap Index Scan on brin_minmax_int2 (actual rows=3460 loops=1)
               Index Cond: (dataset = '{100}'::smallint[])
               Buffers: shared hit=3
 Planning:
   Buffers: shared hit=1
 Planning Time: 0.064 ms
 Execution Time: 0.574 ms
(14 rows)

SELECT count(*) FROM test WHERE dataset = '{100}'::int2[];
 count
-------
     0
(1 row)

EXPLAIN (costs off, analyze, verbose off, buffers, timing off)
	SELECT count(*) FROM test WHERE dataset = '{68,73,74,77,79,80,84,85,88,89,90,94,99,100,101}'::int2[];
                                              QUERY PLAN
------------------------------------------------------------------------------------------------------
 Aggregate (actual rows=1 loops=1)
   Buffers: shared hit=47
   ->  Bitmap Heap Scan on test (actual rows=1 loops=1)
         Recheck Cond: (dataset = '{68,73,74,77,79,80,84,85,88,89,90,94,99,100,101}'::smallint[])
         Rows Removed by Index Recheck: 43
         Heap Blocks: lossy=44
         Buffers: shared hit=47
         ->  Bitmap Index Scan on brin_minmax_int2 (actual rows=440 loops=1)
               Index Cond: (dataset = '{68,73,74,77,79,80,84,85,88,89,90,94,99,100,101}'::smallint[])
               Buffers: shared hit=3
 Planning:
   Buffers: shared hit=1
 Planning Time: 0.026 ms
 Execution Time: 0.511 ms
(14 rows)

SELECT count(*) FROM test WHERE dataset = '{68,73,74,77,79,80,84,85,88,89,90,94,99,100,101}'::int2[];
 count
-------
     1
(1 row)

